### PR TITLE
Merge Development

### PR DESCRIPTION
**Pull request by RidwanHamid501**
Merge Development branch into main

### CODE REVIEW

File: .github/workflows/development.yml [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+          total_count=${{ env.TOTAL_COUNT }}
+          color="green"
+          if [ "$pass_count" -lt "$total_count" ]; then
+            color="yellow"
+          fi
+          if [ "$pass_count" -eq "0" ]; then
+            color="red"
+          fi
+          echo '<svg xmlns="http://www.w3.org/2000/svg" width="150" height="20" role="img" aria-label="Test Results"><linearGradient id="a" x2="0" y2="100%"><stop offset="0" stop-color="#bbb" stop-opacity=".1"/><stop offset="1" stop-opacity=".1"/></linearGradient><rect rx="3" width="150" height="20" fill="#555"/><rect rx="3" x="90" width="60" height="20" fill="'$color'"/><path fill="'$color'" d="M90 0h4v20h-4z"/><rect rx="3" width="150" height="20" fill="url(#a)"/><g fill="#fff" text-anchor="middle" font-family="Verdana,DejaVu Sans,sans-serif" font-size="11"><text x="45" y="15" fill="#010101" fill-opacity=".3">Tests</text><text x="45" y="14">Tests</text><text x="120" y="15" fill="#010101" fill-opacity=".3">'$pass_count'/'$total_count'</text><text x="120" y="14">'$pass_count'/'$total_count'</text></g></svg>' > .github/badges/test.svg
+
```
Comment (SurinaCS) [inline .github/workflows/development.yml:58]: this creates great badges!

File: clean/clean.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    format='%(asctime)s - %(levelname)s - %(message)s',
+    handlers=[
+        logging.StreamHandler(),
+    ]
+)
+
+S3_CLIENT = boto3.client('s3')
+DAY_LIMIT = 7
+load_dotenv(".env")
+
```
Comment (SurinaCS) [inline clean/clean.py:23]: nice use of global variables

File: clean/clean.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                deleted_files.append(object_key)
+                logging.info("Deleted old object: %s", object_key)
+            else:
+                logging.info("Object is within retention period: %s", object_key)
+
+        return {"status": "Completed", "deleted_files": deleted_files}
+
+    except Exception as e:
+        logging.error("Error during cleanup: %s", e)
+        return {"status": "Failed", "error": str(e)}
```
Comment (SurinaCS) [inline clean/clean.py:63]: error handling here is key!

File: dashboard/dashboard.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                       (ENV["SCHEMA_NAME"],))
+        cursor.execute(query, params)
+        if fetch_one:
+            return cursor.fetchone()
+        conn.commit()
+        conn.close()
+        return None
+    except psycopg2.DatabaseError as e:
+        st.error(f"Database error: {e}")
+        return None
```
Comment (SurinaCS) [inline dashboard/dashboard.py:41]: nice error handling

File: dashboard/dashboard.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+                        "Phone number not found. Registering to database.")
+            st.rerun()
+        else:
+            st.warning("Please enter both your name and phone number.")
+
+    if "verification_error" in st.session_state and st.session_state["verification_error"]:
+        st.error(st.session_state["verification_error"])
+
+
+def topic_submission() -> None:
```
Comment (SurinaCS) [inline dashboard/dashboard.py:139]: great typehints

File: pipeline/api.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    return jsonify({"message": "Topic added successfully", "topic": topic}), 200
+
+
+@app.route("/topics", methods=["GET"])
+def get_topics():
+    return jsonify({"topics": topics}), 200
+
+
+if __name__ == "__main__":
+    app.run(host="0.0.0.0", port=5000, debug=True)
```
Comment (SurinaCS) [inline pipeline/api.py:46]: nice concise

File: pipeline/extract.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+def create_dataframe(topic: list[str]) -> pd.DataFrame:
+    """Main function to extract data and return a DataFrame of the relevant data"""
+    s3 = s3_connection()
+
+    filenames = extract_bluesky_files(s3)
+    matching_file_texts = multi_threading_matching(s3, topic, filenames)
+    if matching_file_texts:
+        return pd.DataFrame(matching_file_texts)
+    raise ValueError("No matches found.")
```
Comment (SurinaCS) [inline pipeline/extract.py:133]: great error handling

File: pipeline/load.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    cursor = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
+    cursor.execute(f"SET SEARCH_PATH TO {ENV["SCHEMA_NAME"]};")
+    cursor.execute(
+        f"""INSERT INTO topic (topic_name)
+        VALUES ('{keyword}')""")
+    conn.commit()
+
+
+def main(keyword: str):
+
```
Comment (SurinaCS) [inline pipeline/load.py:25]: missing module docstring and typehints
```suggestion
def main(keyword: str) -> None:
    """Main function to load environment variables to import data into the database."""
```

File: pipeline/transform.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+                (keyword_lower,)
+            )
+            connection.commit()
+            new_keyword_id = cursor.fetchone()['keywords_id']
+            keyword_map[keyword_lower] = new_keyword_id
+
+    return keyword_map
+
+
+def keyword_matching(cleaned_bluesky_data: pd.DataFrame, keyword_map: dict) -> pd.DataFrame:
```
Comment (SurinaCS) [inline pipeline/transform.py:72]: create typehints

File: upload/upload.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+def format_text(text: str) -> str:
+    """Removes extra lines and whitespaces from text"""
+    # Remove consecutive empty lines
+    text = re.sub(r'\n\s*\n', '\n', text)
+
+    text = text.strip()
+    # Replace multiple spaces with a single space
+    text = re.sub(r'\s+', ' ', text)
+    return text
```
Comment (SurinaCS) [inline upload/upload.py:56]: great use of rds

**Review by SurinaCS — CHANGES_REQUESTED** [review]
just a small change re missing docstring, added as suggestion

File: .github/workflows/development.yml
Comment (Kurt812) [inline .github/workflows/development.yml:1]: CI/CD is useful and important!

File: .gitignore
Comment (Kurt812) [inline .gitignore:1]: Nice use of .gitignore to keep the repo tidy and keep our secrets private.

File: images/ERD-diagram.png
Comment (Kurt812) [inline images/ERD-diagram.png:1]: This image needs to be updated with our latest ERD

File: clean/clean.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+
+def s3_connection() -> connection:
+    """Connects to an S3"""
+    s3 = client("s3", aws_access_key_id=os.environ.get("ACCESS_KEY_ID"),
+                aws_secret_access_key=os.environ.get("SECRET_ACCESS_KEY"))
+    return s3
+
+
+def lambda_handler(event, context):
```
Comment (Kurt812) [inline clean/clean.py:32]: Could this use some typehints?

**Review by SurinaCS — APPROVED** [review]

**Pull request merged**